File: csharp/lesson 3/LongestSubstring.cs
using System;
using System.Collections.Generic;

// https://leetcode.com/problems/longest-substring-without-repeating-characters/
public class Solution
{
    public int LengthOfLongestSubstring(string s)
    {
        int left = 0, right = 0, result = 0;
        var map = new Dictionary<char, int>();

        while (right < s.Length)
        {
            if (map.ContainsKey(s[right]))
            {
                left = Math.Max(left, map[s[right]] + 1);
            }

            map[s[right]] = right;
            result = Math.Max(result, right - left + 1);
            right++;
        }

        return result;
    }
}

File: javascript/lesson 3/longestSubstring.js
// https://leetcode.com/problems/longest-substring-without-repeating-characters

function lengthOfLongestSubstring(s) {
    let left = 0, right = 0, result = 0;
    const map = new Map();

    while (right < s.length) {
        if (map.has(s[right])) {
            left = Math.max(left, map.get(s[right]) + 1);
        }

        map.set(s[right], right);
        result = Math.max(result, right - left + 1);
        right++;
    }

    return result;
}
